pkg/broker/ingress/multi_topic_decouple_sink.go

```diff
@@ -0,0 +1,104 @@
+package ingress
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/pkg/errors"
+	"go.uber.org/zap"
+
+	cev2 "github.com/cloudevents/sdk-go/v2"
+	cecontext "github.com/cloudevents/sdk-go/v2/context"
+	"github.com/cloudevents/sdk-go/v2/protocol"
+	"github.com/cloudevents/sdk-go/v2/protocol/pubsub"
+	"github.com/google/knative-gcp/pkg/broker/config"
+	"github.com/google/knative-gcp/pkg/broker/config/volume"
+	"knative.dev/eventing/pkg/logging"
+)
+
+// NewMultiTopicDecoupleSink creates a new multiTopicDecoupleSink.
+func NewMultiTopicDecoupleSink(ctx context.Context, options ...MultiTopicDecoupleSinkOption) (*multiTopicDecoupleSink, error) {
+	sink := &multiTopicDecoupleSink{
+		logger: logging.FromContext(ctx),
+	}
+
+	for _, opt := range options {
+		opt(sink)
+	}
+
+	// Aplly defaults
+	if sink.client == nil {
+		client, err := newDefaultPubSubClient(ctx)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create pubsub client: %w", err)
```

Comment:
Just for my own knowledge: as I haven't used `errors.Wrap` yet, when do we generally wrap errors vs using `Errorf`?

Reply:
I found `errors.Wrap` to be very convenient as it adds context information by recording a stack trace. 

> // Wrap returns an error annotating err with a stack trace
// at the point Wrap is called, and the supplied message.
// If err is nil, Wrap returns nil.

Reply:
I haven't used wrapping much yet either.  You probaby want to avoid using Wrap() if you're at a level where your code completely understands the error and is capable of given a full description of the error.   This is particularly true if you're in a path where upper layers may want to be able to do differing things depending on the type of the error.  In particular, if you thing the error your function is returning may be something that users will want to detect and deal with(but one of the errors you are wrapping is quite generic), you may want to use your own error type and not wrap.  With unwrapping you can use Is() on the error to match an error chain where any element on the chain unwraps to that error type.  If you have an error path that can return multiple "interesting" error types (sql not found errors for multiple databases, etc), or something generic like EOF (but where the EOF isn't really logically your error) you might just want to use your own type and not wrap the sql error to avoid exposing your internal implementation.

Reply:
Sorry I didn't read the line carefully. You are right I don't need to wrap the pubsusb error here.